src/modules/simulator/simulator_mavlink.cpp

```diff
@@ -1128,32 +1127,93 @@ int Simulator::publish_flow_topic(mavlink_hil_optical_flow_t *flow_mavlink)
 	return OK;
 }
 
-int Simulator::publish_ev_topic(mavlink_vision_position_estimate_t *ev_mavlink)
+template<typename T>
+int Simulator::publish_odometry_topic(T *msg)
 {
 	uint64_t timestamp = hrt_absolute_time();
 
-	struct vehicle_local_position_s vision_position = {};
+	struct vehicle_odometry_s odom = {};
+
+	odom.timestamp = timestamp;
+
+	if (msg->msgid == MAVLINK_MSG_ID_ODOMETRY) {
+		mavlink_odometry_t odom_msg;
+		mavlink_msg_odometry_decode(msg, &odom_msg);
+
+		/* Dcm rotation matrix from body frame to local NED frame */
+		matrix::Dcm<float> Rbl;
+
+		/* since odom.child_frame_id == MAV_FRAME_BODY_FRD, WRT to estimated vehicle body-fixed frame */
+		/* get quaternion from the msg quaternion itself and build DCM matrix from it */
+		/* No need to transform the covariance matrices since the non-diagonal values are all zero */
+		Rbl = matrix::Dcm<float>(matrix::Quatf(odom_msg.q)).I();
+
+		/* the linear velocities needs to be transformed to the local NED frame */
+		matrix::Vector3<float> linvel_local(Rbl * matrix::Vector3<float>(odom_msg.vx, odom_msg.vy, odom_msg.vz));
+
+		/* The position in the local NED frame */
+		odom.x = odom_msg.x;
+		odom.y = odom_msg.y;
+		odom.z = odom_msg.z;
+		/* The quaternion of the ODOMETRY msg represents a rotation from
+		 * NED earth/local frame to XYZ body frame */
+		matrix::Quatf q(odom_msg.q[0], odom_msg.q[1], odom_msg.q[2], odom_msg.q[3]);
+		q.copyTo(odom.q);
 
-	vision_position.timestamp = timestamp;
-	vision_position.x = ev_mavlink->x;
-	vision_position.y = ev_mavlink->y;
-	vision_position.z = ev_mavlink->z;
+		/* The pose covariance URT */
+		for (size_t i = 0; i < 21; i++) {
```

Comment:
Should use `sizeof(odom.pose_covariance) / sizeof(odom.pose_covariance[0])` instead of the hardcoded 21

Reply:
sure.